### PR TITLE
Remove `handleEpochBoundary` and move shuffling caches to post next slot cache fill

### PR DESCRIPTION
**Pull request by terencechain**
Review and merge #12484 first

There are a few issues concerning how Prysm updates shuffling caches (attester committee + proposer)
- Missing skip slot case. If slot 31 is missed, then the next epoch slot 0/1/2 cache usages will be slow 
- Independent of next slot cache update. This means we are doing the work twice 
- The usage of `s.nextEpochBoundarySlot` is not reorg resistant 

This PR addresses the above by removing `handleEpochBoundary` and moving the shuffling caches update function next to updating the next slot cache under a go routine. It also add an `updateShufflingCachesOnReorg` as soon as reorg is detected in `onBlock`. `updateShufflingCachesOnReorg` fills the cache based on head state

### CODE REVIEW

File: beacon-chain/blockchain/process_block.go [hunk 1]
```diff
@@ -482,52 +483,6 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []interfaces.ReadOnlySi
 	return s.saveHeadNoDB(ctx, lastB, lastBR, preState)
 }
 
-// Epoch boundary bookkeeping such as logging epoch summaries.
-func (s *Service) handleEpochBoundary(ctx context.Context, postState state.BeaconState) error {
```
Comment (potuz) [inline beacon-chain/blockchain/process_block.go:486]: The removal of this function should be in the description of the PR and most probably on the title as well.

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -720,3 +681,31 @@ func (s *Service) lateBlockTasks(ctx context.Context) {
...
+
+// updateShufflingCache updates the following caches with respect the validator shuffling.
+// 1.) committee cache
+// 2.) proposer indices cache
+// no-op if the last cached next slot state is less than next boundary slot.
+func (s *Service) updateShufflingCaches(ctx context.Context) error {
+	ctx, span := trace.StartSpan(ctx, "transition.updateShufflingCaches")
+	defer span.End()
+
+	transition.LastCachedState()
```
Comment (potuz) [inline beacon-chain/blockchain/process_block.go:693]: remove this?

File: beacon-chain/blockchain/process_block.go [hunk 3]
```diff
@@ -720,3 +681,34 @@ func (s *Service) lateBlockTasks(ctx context.Context) {
...
+func (s *Service) updateShufflingCaches(ctx context.Context) error {
+	ctx, span := trace.StartSpan(ctx, "transition.updateShufflingCaches")
+	defer span.End()
+
+	var err error
+	lastCachedSlot := transition.LastCachedStateSlot()
+	lastCachedEpoch := slots.ToEpoch(lastCachedSlot)
+
+	// Allow cache update if one of the two condition meet
+	// 1.) last cached is great or equal than next epoch boundary slot
```
Comment (rkapka) [inline beacon-chain/blockchain/process_block.go:697]:
```suggestion
	// 1.) last cached state slot is great or equal than next epoch boundary slot
```

File: beacon-chain/blockchain/process_block.go [hunk 4]
```diff
@@ -696,6 +651,12 @@ func (s *Service) lateBlockTasks(ctx context.Context) {
 		log.WithError(err).Debug("could not update next slot state cache")
 	}
 
+	go func() {
+		if err := s.updateShufflingCaches(ctx); err != nil {
```
Comment (rkapka) [inline beacon-chain/blockchain/process_block.go:655]: Why do we do this both here and in `onBlock`? `insertBlockToForkchoiceStore` is called inside `onBlock`, meaning that shuffling will be called twice.

Comment (terencechain) [reply beacon-chain/blockchain/process_block.go:655]: this is not part of `onBlock`

Comment (rkapka) [reply beacon-chain/blockchain/process_block.go:661]: oops, I misread it on Github, sorry

File: beacon-chain/blockchain/process_block.go [hunk 5]
```diff
@@ -194,6 +194,12 @@ func (s *Service) onBlock(ctx context.Context, signed interfaces.ReadOnlySignedB
 		log.WithError(err).Warn("Could not update head")
 	}
 	if blockRoot != headRoot {
+
```
Comment (rkapka) [inline beacon-chain/blockchain/process_block.go:197]:
```suggestion
```

**Comment by terencechain** [comment]
This is still relevant but I still start another PR since much has changed

**Pull request closed without merging**